### ISSUE TITLE
Leaking memory with custom data type and reliable publisher, best_effort subscriber

### ISSUE THREAD

**Issue by jespersmith** [eProsima/Fast-RTPS#214]
I'm using a custom data type with the pub/sub interface (see for implementation at  https://github.com/ihmcrobotics/ihmc-pub-sub/blob/develop/ihmc-pub-sub/cppsrc/FastRTPS/rawtopicdatatype.cpp). The bug does not seem the appear with a generated message (but that could be a red herring)

When I use this type to publish data on a reliable publisher with best effort subscriber a large amount of memory gets allocated internally in FastRTPS, in the following trace from valgrind --tool=massif:

```
->26.82% (2,804,048B) 0x1D8A7C: std::_Rb_tree_iterator<eprosima::fastrtps::rtps::ChangeForReader_t> std::_Rb_tree<eprosima::fastrtps::rtps::ChangeForReader_t, eprosima::fastrtps::rtps::ChangeForReader_t, std::_Identity<eprosima::fastrtps::rtps::ChangeForReader_t>, eprosima::fastrtps::rtps::ChangeForReaderCmp, std::allocator<eprosima::fastrtps::rtps::ChangeForReader_t> >::_M_insert_<eprosima::fastrtps::rtps::ChangeForReader_t const&, std::_Rb_tree<eprosima::fastrtps::rtps::ChangeForReader_t, eprosima::fastrtps::rtps::ChangeForReader_t, std::_Identity<eprosima::fastrtps::rtps::ChangeForReader_t>, eprosima::fastrtps::rtps::ChangeForReaderCmp, std::allocator<eprosima::fastrtps::rtps::ChangeForReader_t> >::_Alloc_node>(std::_Rb_tree_node_base*, std::_Rb_tree_node_base*, eprosima::fastrtps::rtps::ChangeForReader_t const&, std::_Rb_tree<eprosima::fastrtps::rtps::ChangeForReader_t, eprosima::fastrtps::rtps::ChangeForReader_t, std::_Identity<eprosima::fastrtps::rtps::ChangeForReader_t>, eprosima::fastrtps::rtps::ChangeForReaderCmp, std::allocator<eprosima::fastrtps::rtps::ChangeForReader_t> >::_Alloc_node&) [clone .isra.244] (new_allocator.h:111)
| ->26.81% (2,803,840B) 0x1DA68F: eprosima::fastrtps::rtps::ReaderProxy::setNotValid(eprosima::fastrtps::rtps::CacheChange_t*) (stl_tree.h:1026)
| | ->26.81% (2,803,840B) 0x1CDA81: eprosima::fastrtps::rtps::StatefulWriter::change_removed_by_history(eprosima::fastrtps::rtps::CacheChange_t*) (StatefulWriter.cpp:200)
| |   ->26.81% (2,803,840B) 0x1E2BAD: eprosima::fastrtps::rtps::WriterHistory::remove_change(eprosima::fastrtps::rtps::CacheChange_t*) (WriterHistory.cpp:118)
| |     ->26.81% (2,803,840B) 0x193CE3: eprosima::fastrtps::PublisherHistory::remove_change_pub(eprosima::fastrtps::rtps::CacheChange_t*, __gnu_cxx::__normal_iterator<std::pair<eprosima::fastrtps::rtps::InstanceHandle_t, std::vector<eprosima::fastrtps::rtps::CacheChange_t*, std::allocator<eprosima::fastrtps::rtps::CacheChange_t*> > >*, std::vector<std::pair<eprosima::fastrtps::rtps::InstanceHandle_t, std::vector<eprosima::fastrtps::rtps::CacheChange_t*, std::allocator<eprosima::fastrtps::rtps::CacheChange_t*> > >, std::allocator<std::pair<eprosima::fastrtps::rtps::InstanceHandle_t, std::vector<eprosima::fastrtps::rtps::CacheChange_t*, std::allocator<eprosima::fastrtps::rtps::CacheChange_t*> > > > > >*) (PublisherHistory.cpp:243)
| |       ->26.81% (2,803,840B) 0x1DF891: eprosima::fastrtps::rtps::WriterHistory::remove_min_change() (WriterHistory.cpp:216)
| |         ->26.81% (2,803,840B) 0x194E9B: eprosima::fastrtps::PublisherHistory::add_pub_change(eprosima::fastrtps::rtps::CacheChange_t*, eprosima::fastrtps::rtps::WriteParams&, std::unique_lock<std::recursive_mutex>&) (PublisherHistory.cpp:76)
| |           ->26.81% (2,803,840B) 0x18C98D: eprosima::fastrtps::PublisherImpl::create_new_change_with_params(eprosima::fastrtps::rtps::ChangeKind_t, void*, eprosima::fastrtps::rtps::WriteParams&) (PublisherImpl.cpp:167)

```

**Comment by MiguelCompany** [comment]
We have made some changes (commits f8e3abf1c2790457c3e995a18b28a029c3d19793 and bb42407a0d304d01057b80b20cf0fcbd9e6e8a5d) that I think are related to your problem.

Could you please check if using the current HEAD of master solves your problem?

**Comment by MiguelCompany** [comment]
I see it now. It's [ReaderProxy::setNotValid](https://github.com/eProsima/Fast-RTPS/blob/54f563a697e7de5e3c924a99fa1b2ce3877f69e7/src/cpp/rtps/writer/ReaderProxy.cpp#L319) removing a ChangeForReader_t from a set and then inserting a modified version of it again. We will work on this, but I cannot give you an estimated date for it.

Thanks for finding this issue.

**Comment by MiguelCompany** [comment]
Closing this issue, as this should be solved on release 1.7.0. Feel free to reopen it if you're still having problems.
